tests/system/utils.py

```diff
@@ -143,3 +143,75 @@ def convert_pandas_dtypes(df: pd.DataFrame, bytes_col: bool):
     df["numeric_col"] = df["numeric_col"].apply(
         lambda value: decimal.Decimal(str(value)) if value else None  # type: ignore
     )
+
+
+def assert_pandas_df_equal_pca_components(actual, expected, **kwargs):
+    """Compare two pandas dataframes representing PCA components. The columns
+    required to be present in the dataframes are:
+        numerical_value: numeric,
+        categorical_value: List[object(category, value)]
+
+    Args:
+        actual: Actual Pandas DataFrame
+
+        expected: Expected Pandas DataFrame
+
+        kwargs: kwargs to use in `pandas.testing.assert_series_equal` per column
+    """
+    # Compare the index, columns and values separately, as the polarity of the
+    # PCA vectors can be arbitrary
+    pd.testing.assert_index_equal(
+        actual.index, expected.index.astype(actual.index.dtype)
+    )  # dtype agnostic index comparison
+    pd.testing.assert_index_equal(actual.columns, expected.columns)
+    for column in expected.columns:
+        try:
+            pd.testing.assert_series_equal(actual[column], expected[column], **kwargs)
+        except AssertionError:
+            if column not in {"numerical_value", "categorical_value"}:
+                raise
+
+            # Allow for sign difference per numeric/categorical column
+            if column == "numerical_value":
+                actual_ = -actual[column]
+                expected_ = expected[column]
+            else:
+                # In this column each element is an array of objects, where the
+                # object has attributes "category" and "value". For the sake of
+                # comparison let's normalize by flipping the polarity of "value".
+                def normalize_array_of_objects(arr, reverse_polarity=False):
+                    newarr = []
+                    for element in arr:
+                        newelement = dict(element)
+                        if reverse_polarity:
+                            newelement["value"] = -newelement["value"]
+                        newarr.append(newelement)
+                    return sorted(newarr, key=lambda d: d["category"])
+
+                actual_ = actual[column].apply(normalize_array_of_objects, args=(True,))
+                expected_ = expected[column].apply(normalize_array_of_objects)
+
+            pd.testing.assert_series_equal(actual_, expected_, **kwargs)
+
+
+def assert_pandas_df_equal_pca(actual, expected, **kwargs):
```

Comment:
Is it that this can be replace by assert_pandas_df_equal_pca_components?

Reply:
Let me improve the docstring to reflect that `assert_pandas_df_equal_pca_components` compares the principle components themselves, while `assert_pandas_df_equal_pca` compares the projection of rows in the PC dimensions.